src/app/bdx/DiagnosticLogsBDXTransferHandler.cpp

```diff
@@ -0,0 +1,275 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/core/CHIPConfig.h>
+
+#if CHIP_CONFIG_ENABLE_BDX_LOG_TRANSFER
+#include "DiagnosticLogsBDXTransferHandler.h"
+#include <app/clusters/diagnostic-logs-server/diagnostic-logs-server.h>
+#include <messaging/ExchangeMgr.h>
+#include <platform/CHIPDeviceLayer.h>
+
+using namespace chip;
+using namespace app;
+using namespace bdx;
+using namespace app::Clusters::DiagnosticLogs;
+
+// BDX transfer session polling interval.
+constexpr System::Clock::Timeout kBdxPollIntervalMs = System::Clock::Milliseconds32(50);
+
+// Timeout for the BDX transfer session
+constexpr System::Clock::Timeout kBdxTimeout = System::Clock::Seconds16(5 * 60);
+
+constexpr uint16_t kBdxMaxBlockSize = 1024;
+
+StatusCode GetBdxStatusCodeFromChipError(CHIP_ERROR err)
+{
+    if (err == CHIP_ERROR_INCORRECT_STATE)
+    {
+        return StatusCode::kUnexpectedMessage;
+    }
+    if (err == CHIP_ERROR_INVALID_ARGUMENT)
+    {
+        return StatusCode::kBadMessageContents;
+    }
+    return StatusCode::kUnknown;
+}
+
+CHIP_ERROR DiagnosticLogsBDXTransferHandler::InitializeTransfer(Messaging::ExchangeManager * exchangeMgr,
+                                                                const SessionHandle sessionHandle, FabricIndex fabricIndex,
+                                                                NodeId peerNodeId, LogProviderDelegate * delegate,
+                                                                IntentEnum intent, CharSpan fileDesignator)
+{
+    if (mInitialized)
+    {
+        // Return busy if we are in a transfer session already with another node.
+        VerifyOrReturnError(mFabricIndex.Value() == fabricIndex && mPeerNodeId.Value() == peerNodeId, CHIP_ERROR_BUSY);
+
+        // Reset stale connection from the same Node if exists.
```

Comment:
if the fabric index and the node id in the above block of code matches the stored fabric index and node id. i will add an explicit comment

Reply:
No, my point is: when would that ever happen?

Reply:
actually it won't. This is totally my bad. Since we are creating a new object for each transfer, it's not a problem. This makes me think handling busy will be trickier since busy will not work this way. I haven't added a test case for busy yet but i should to make sure the code works.